beacon-chain/core/validators/validator.go

```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
+//             validator.activation_eligibility_epoch = get_current_epoch(state)
+//         if is_active_validator(validator, get_current_epoch(state)) and validator.effective_balance <= EJECTION_BALANCE:
+//             initiate_validator_exit(state, index)
+//     # Queue validators eligible for activation and not dequeued for activation prior to finalized epoch
+//     activation_queue = sorted([
+//         index for index, validator in enumerate(state.validator_registry) if
+//         validator.activation_eligibility_epoch != FAR_FUTURE_EPOCH and
+//         validator.activation_epoch >= get_delayed_activation_exit_epoch(state.finalized_epoch)
+//     ], key=lambda index: state.validator_registry[index].activation_eligibility_epoch)
+//     # Dequeued validators for activation up to churn limit (without resetting activation epoch)
+//     for index in activation_queue[:get_churn_limit(state)]:
+//         validator = state.validator_registry[index]
+//         if validator.activation_epoch == FAR_FUTURE_EPOCH:
+//             validator.activation_epoch = get_delayed_activation_exit_epoch(get_current_epoch(state))
+func UpdateRegistry(state *pb.BeaconState) *pb.BeaconState {
 	currentEpoch := helpers.CurrentEpoch(state)
-	updatedEpoch := helpers.DelayedActivationExitEpoch(currentEpoch)
-	activeValidatorIndices := helpers.ActiveValidatorIndices(state.ValidatorRegistry, currentEpoch)
-	totalBalance := helpers.TotalBalance(state, activeValidatorIndices)
-
-	// The maximum balance churn in Gwei (for deposits and exits separately).
-	maxBalChurn := maxBalanceChurn(totalBalance)
-
-	var balChurn uint64
-	var err error
 	vStore.Lock()
 	defer vStore.Unlock()
 	for idx, validator := range state.ValidatorRegistry {
 		// Activate validators within the allowable balance churn.
-		if validator.ActivationEpoch == params.BeaconConfig().FarFutureEpoch &&
-			state.Balances[idx] >= params.BeaconConfig().MaxDepositAmount &&
-			!helpers.IsActiveValidator(validator, currentEpoch) {
-			balChurn += helpers.EffectiveBalance(state, uint64(idx))
-			log.WithFields(logrus.Fields{
-				"index":               idx,
-				"currentBalanceChurn": balChurn,
-				"maxBalanceChurn":     maxBalChurn,
-				"currentEpoch":        currentEpoch - params.BeaconConfig().GenesisEpoch,
-			}).Info("Attempting to activate validator")
-
-			if balChurn > maxBalChurn {
-				break
-			}
-			state, err = ActivateValidator(state, uint64(idx), false)
-			if err != nil {
-				return nil, fmt.Errorf("could not activate validator %d: %v", idx, err)
-			}
-			vStore.activatedValidators[updatedEpoch] =
-				append(vStore.activatedValidators[updatedEpoch], uint64(idx))
+		if validator.ActivationEligibilityEpoch == params.BeaconConfig().FarFutureEpoch &&
+			validator.EffectiveBalance >= params.BeaconConfig().MaxEffectiveBalance {
+			validator.ActivationEligibilityEpoch = currentEpoch
```

Comment:
This code path is not tested, can we add test to cover this line?

Reply:
adde tests, thanks